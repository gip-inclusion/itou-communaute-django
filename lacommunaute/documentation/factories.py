import factory
from faker import Faker

from lacommunaute.documentation.models import Category, Document, DocumentRating


faker = Faker()


class AbstractDocumentationFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=100)
    short_description = factory.Faker("sentence", nb_words=10)
    image = factory.django.ImageField(filename="banner.jpg")


class CategoryFactory(AbstractDocumentationFactory):
    class Meta:
        model = Category

    class Params:
        for_snapshot = factory.Trait(
            name="Test Category", description="Test description", short_description="Test short description"
        )


class DocumentFactory(AbstractDocumentationFactory):
    category = factory.SubFactory(CategoryFactory)

    class Meta:
        model = Document

    class Params:
        for_snapshot = factory.Trait(
            name="Test Document",
            description="Test description",
            short_description="Test short description",
            category=factory.SubFactory(CategoryFactory, for_snapshot=True),
        )

    @factory.post_generation
    def with_tags(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        if isinstance(extracted, list):
            for tag in extracted:
                self.tags.add(tag)


class DocumentRatingFactory(factory.django.DjangoModelFactory):
    document = factory.SubFactory(DocumentFactory)
    rating = factory.Faker("random_int", min=1, max=5)

    class Meta:
        model = DocumentRating
        skip_postgeneration_save = True

    @factory.post_generation
    def set_created(self, create, extracted, **kwargs):
        if extracted:
            self.created = extracted
            self.save()
