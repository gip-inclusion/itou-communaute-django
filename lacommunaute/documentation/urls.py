from django.urls import path

from lacommunaute.documentation.views import (
    CategoryCreateView,
    CategoryDetailView,
    CategoryListView,
    CategoryUpdateView,
    DocumentDetailView,
)


app_name = "documentation"


urlpatterns = [
    path("", CategoryListView.as_view(), name="category_list"),
    path("<str:slug>-<int:pk>/", CategoryDetailView.as_view(), name="category_detail"),
    path("category/create/", CategoryCreateView.as_view(), name="category_create"),
    path("<str:slug>-<int:pk>/update/", CategoryUpdateView.as_view(), name="category_update"),
    path("<int:category_pk>/<str:slug>-<int:pk>/", DocumentDetailView.as_view(), name="document_detail"),
]
