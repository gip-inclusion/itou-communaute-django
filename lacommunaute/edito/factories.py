import factory
from faker import Faker

from lacommunaute.edito.models import Edito
from lacommunaute.users.factories import UserFactory


faker = Faker()


class EditoFactory(factory.django.DjangoModelFactory):
    poster = factory.SubFactory(UserFactory)
    title = faker.sentence()
    content = faker.text()
    url = faker.url()
    link_title = faker.sentence()

    class Meta:
        model = Edito

    class Params:
        for_snapshot = factory.Trait(
            title="Snapshot title",
            content="Snapshot *content*",
            url="https://example.com",
            link_title="Snapshot link title",
        )
