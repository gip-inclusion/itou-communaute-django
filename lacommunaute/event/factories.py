import datetime

import factory
from factory import fuzzy
from faker import Faker

from lacommunaute.event.models import Event
from lacommunaute.users.factories import UserFactory


faker = Faker()


class EventFactory(factory.django.DjangoModelFactory):
    poster = factory.SubFactory(UserFactory)
    name = factory.Faker("name")
    date = fuzzy.FuzzyDate(datetime.date(2023, 5, 17), datetime.date(2023, 7, 16))
    time = datetime.time(10)
    end_date = factory.LazyAttribute(lambda o: o.date)
    end_time = factory.LazyAttribute(lambda o: o.time.replace(hour=(o.time.hour + 1) % 24))
    location = faker.url()
    description = factory.LazyAttribute(lambda t: faker.text(max_nb_chars=200))

    class Meta:
        model = Event
