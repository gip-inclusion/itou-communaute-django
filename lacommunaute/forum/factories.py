import factory
from machina.core.db.models import get_model
from machina.test.factories.forum import ForumFactory as BaseForumFactory

from lacommunaute.forum.models import Forum
from lacommunaute.users.factories import GroupFactory


ForumPermission = get_model("forum_permission", "ForumPermission")
UserForumPermission = get_model("forum_permission", "UserForumPermission")


class ForumFactory(BaseForumFactory):
    type = Forum.FORUM_POST
    members_group = factory.SubFactory(GroupFactory, name=factory.SelfAttribute("..name"))
    name = factory.Faker("name")
    description = factory.Faker("sentence", nb_words=10)

    @factory.post_generation
    def with_public_perms(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        perms = [
            "can_see_forum",
            "can_read_forum",
            "can_start_new_topics",
            "can_reply_to_topics",
            "can_edit_own_posts",
            "can_delete_own_posts",
            "can_post_without_approval",
        ]

        anonymous_authorized_perms = [
            UserForumPermission(
                anonymous_user=True, authenticated_user=False, permission=permission, has_perm=True, forum=self
            )
            for permission in ForumPermission.objects.filter(codename__in=perms)
        ]

        authentified_authorized_perms = [
            UserForumPermission(
                anonymous_user=False, authenticated_user=True, permission=permission, has_perm=True, forum=self
            )
            for permission in ForumPermission.objects.filter(codename__in=perms)
        ]
        UserForumPermission.objects.bulk_create(anonymous_authorized_perms + authentified_authorized_perms)
