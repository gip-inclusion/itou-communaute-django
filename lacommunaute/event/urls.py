from django.urls import path

from lacommunaute.event.views import (
    EventCreateView,
    EventDeleteView,
    EventDetailView,
    EventListView,
    EventMonthArchiveView,
    EventUpdateView,
)


app_name = "event"


urlpatterns = [
    path("create/", EventCreateView.as_view(), name="create"),
    path("<int:pk>/", EventDetailView.as_view(), name="detail"),
    path("<int:pk>/update/", EventUpdateView.as_view(), name="update"),
    path("<int:pk>/delete/", EventDeleteView.as_view(), name="delete"),
    path("myevents/", EventListView.as_view(), name="myevents"),
    path("events/<int:year>/<int:month>/", EventMonthArchiveView.as_view(), name="month"),
    path("events/", EventMonthArchiveView.as_view(), name="current"),
]
