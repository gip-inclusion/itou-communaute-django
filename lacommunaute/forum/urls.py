from django.urls import path
from machina.apps.forum.views import IndexView

from lacommunaute.forum.views import CategoryForumListView, ForumView


app_name = "forum_extension"


urlpatterns = [
    path("forum/<str:slug>-<int:pk>/", ForumView.as_view(), name="forum"),
    path("forum/documentation/", CategoryForumListView.as_view(), name="documentation"),
    path("forum/", IndexView.as_view(), name="index"),
]
