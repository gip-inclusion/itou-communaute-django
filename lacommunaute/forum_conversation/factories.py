import factory
from machina.test.factories.conversation import PostFactory as BasePostFactory, TopicFactory as BaseTopicFactory

from lacommunaute.forum.factories import ForumFactory
from lacommunaute.forum_conversation.forum_polls.factories import TopicPollVoteFactory
from lacommunaute.forum_conversation.models import CertifiedPost, Topic
from lacommunaute.users.factories import UserFactory


class PostFactory(BasePostFactory):
    subject = factory.Faker("sentence", nb_words=5)
    content = factory.Faker("sentence", nb_words=40)
    poster = factory.SubFactory(UserFactory)


class CertifiedPostFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = CertifiedPost


class TopicFactory(BaseTopicFactory):
    forum = factory.SubFactory(ForumFactory)
    poster = factory.SubFactory(UserFactory)
    subject = factory.Faker("sentence", nb_words=5)
    type = Topic.TOPIC_POST

    class Params:
        with_post = factory.Trait(
            post=factory.RelatedFactory(
                PostFactory,
                factory_related_name="topic",
                poster=factory.SelfAttribute("topic.poster"),
            )
        )

    @factory.post_generation
    def with_like(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        self.likers.add(self.poster.id)

    @factory.post_generation
    def with_poll_vote(self, create, extracted, **kwargs):
        if not create or not extracted:
            return

        TopicPollVoteFactory(poll_option__poll__topic=self, voter=self.poster)

    @factory.post_generation
    def with_certified_post(self, create, extracted, **kwargs):
        if not create or not extracted:
            return
        PostFactory(topic=self, poster=self.poster)
        CertifiedPostFactory(topic=self, post=PostFactory(topic=self, poster=self.poster), user=self.poster)
