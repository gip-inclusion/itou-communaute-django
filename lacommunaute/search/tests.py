import urllib.parse

import pytest
from django.db import connection
from django.urls import reverse
from pytest_django.asserts import assertContains, assertNotContains

from lacommunaute.forum.factories import ForumFactory
from lacommunaute.forum_conversation.factories import PostFactory, TopicFactory
from lacommunaute.users.factories import UserFactory
from lacommunaute.utils.testing import parse_response_to_soup


@pytest.fixture(name="search_url")
def search_url_fixture():
    return reverse("search:index")


@pytest.fixture(name="forums")
def forums_fixture():
    forum1 = ForumFactory.create(
        name="Le PASS IAE",
        description="Tout savoir sur le PASS IAE, l’insertion par l’activité économique, et plus encore !",
        short_description="Tout savoir sur le PASS IAE",
    )
    topic1 = TopicFactory(forum=forum1, subject="Obtention d’un PASS IAE")
    PostFactory(
        topic=topic1,
        subject="Qui contacter ?",
        content="L’équipe des emplois de l’inclusion",
    )
    forum2 = ForumFactory.create(
        name="Administrateur de structure",
        description="Gérer votre structure, ses membres, sa localisation et ses fiches de postes.",
        short_description="Gestion de la structure",
    )
    topic2 = TopicFactory(forum=forum2, subject="Inviter un collaborateur")
    PostFactory(
        topic=topic2,
        subject="Inviter ses collaborateurs par email",
        content="Listez les emails de vos collaborateurs, ils recevront un email avec les instructions.",
    )
    refresh_search_index()
    return [forum1, forum2]


@pytest.fixture(name="topics")
def topics_fixture():
    forum = ForumFactory.create(
        name="Le rôle de prescripteur",
        description="Explication du rôle de prescripteur, un acteur clé de l’insertion.",
        short_description="Détails du rôle de prescripteur",
    )
    topic1 = TopicFactory(forum=forum, subject="Qui sont les prescripteurs ?")
    PostFactory(
        topic=topic1,
        subject="Les missions locales",
        content="La mission locale est un espace d'intervention au service des jeunes.",
    )
    topic2 = TopicFactory(forum=forum, subject="Habilitation des prescripteurs")
    PostFactory(
        topic=topic2,
        subject="Demander une habilitation",
        content="La demande d’habilitation se fait auprès du préfet.",
    )
    refresh_search_index()
    return [topic1, topic2]


def refresh_search_index():
    with connection.cursor() as c:
        c.execute("REFRESH MATERIALIZED VIEW search_commonindex;")


def test_search_on_post(client, db, search_url, topics):
    query = ["au", "service", "des", "jeunes"]
    response = client.get(search_url, {"q": " ".join(query)})

    assertNotContains(response, "Demander une habilitation")
    for word in ["service", "jeunes"]:  # Stop words are ignored, thus not highlighted.
        assertContains(response, f'<span class="highlighted">{word}</span>')


def test_search_on_forum(client, db, search_url, forums):
    query = ["Tout", "savoir", "sur"]
    response = client.get(search_url, {"q": " ".join(query)})

    assertNotContains(response, forums[1].description.raw)
    for word in ["Tout", "savoir"]:  # Stop words are ignored, thus not highlighted.
        assertContains(response, f'<span class="highlighted">{word}</span>')


def test_search_with_no_query(client, db, search_url):
    ForumFactory()
    response = client.get(search_url)
    assertContains(response, '<input type="search" name="q"')


def test_empty_search(client, db, search_url):
    ForumFactory()
    response = client.get(search_url, {"q": ""})
    assertContains(response, '<input type="search" name="q"')


def test_search_with_no_results(client, db, search_url):
    ForumFactory()
    response = client.get(search_url, {"q": "test"})
    assertContains(response, "Aucun résultat")


def test_search_with_non_unicode_characters(client, db, search_url):
    ForumFactory()
    encoded_char = urllib.parse.quote("\x1f")
    response = client.get(search_url, {"q": encoded_char})
    assertContains(response, "Aucun résultat")


def test_search_on_post_model_only(client, db, search_url, topics, forums):
    datas = {"m": "TOPIC"}

    query = ["La", "mission", "locale", "est"]
    datas["q"] = " ".join(query)

    response = client.get(search_url, datas)
    for word in ["mission", "locale"]:  # Stop words are ignored, thus not highlighted.
        assertContains(response, f'<span class="highlighted">{word}</span>')

    query = ["Tout", "savoir", "sur"]
    datas["q"] = " ".join(query)

    response = client.get(search_url, datas)
    for word in ["Tout", "savoir"]:  # Stop words are ignored, thus not highlighted.
        assertNotContains(response, f'<span class="highlighted">{word}</span>')


def test_search_on_forum_model_only(client, db, search_url, topics, forums):
    datas = {"m": "FORUM"}

    query = ["La", "mission", "locale", "est"]
    datas["q"] = " ".join(query)

    response = client.get(search_url, datas)
    for word in ["mission", "locale"]:  # Stop words are ignored, thus not highlighted.
        assertNotContains(response, f'<span class="highlighted">{word}</span>')

    query = ["Tout", "savoir", "sur"]
    datas["q"] = " ".join(query)

    response = client.get(search_url, datas)
    for word in ["Tout", "savoir"]:  # Stop words are ignored, thus not highlighted.
        assertContains(response, f'<span class="highlighted">{word}</span>')


def test_search_on_both_models(client, db, search_url, topics, forums):
    datas = {"m": "all"}
    query = ["La", "mission", "locale", "est"]
    datas["q"] = " ".join(query)

    response = client.get(search_url, datas)
    for word in ["mission", "locale"]:  # Stop words are ignored, thus not highlighted.
        assertContains(response, f'<span class="highlighted">{word}</span>')

    query = ["Tout", "savoir", "sur", "le"]
    datas["q"] = " ".join(query)

    response = client.get(search_url, datas)
    for word in ["Tout", "savoir"]:  # Stop words are ignored, thus not highlighted.
        assertContains(response, f'<span class="highlighted">{word}</span>')


def test_unapproved_post_is_exclude(client, db, search_url):
    forum = ForumFactory.create(
        name="Le PASS IAE",
        description="Tout savoir sur le PASS IAE, l’insertion par l’activité économique, et plus encore !",
        short_description="Tout savoir sur le PASS IAE",
    )
    topic = TopicFactory(forum=forum, subject="Obtention d’un PASS IAE")
    PostFactory(
        topic=topic,
        approved=False,  # Not approved, should not appear in search results.
        subject="Qui contacter ?",
        content="L’équipe des emplois de l’inclusion",
    )
    refresh_search_index()
    response = client.get(search_url, {"q": "emplois"})
    assertContains(response, "Aucun résultat")


@pytest.mark.parametrize(
    "user,snapshot_name",
    [
        (None, "anonymous"),
        (lambda: UserFactory(username="123"), "proconnected"),
    ],
)
def test_extra_context(client, db, user, snapshot_name, search_url, snapshot):
    forum = ForumFactory()
    if user:
        client.force_login(user())

    response = client.get(search_url)
    content = parse_response_to_soup(response, selector="main")
    assert str(content) == snapshot(name="no_query")

    datas = {"m": "TOPIC", "q": " ".join(["Bubba", "Gump", "Shrimp", "Co."])}
    response = client.get(search_url, datas)
    content = parse_response_to_soup(
        response, selector="main", replace_in_href=[(f"{forum.slug}-{forum.pk}", "forrest-gump-PK")]
    )
    assert str(content) == snapshot(name=snapshot_name)


def test_search_all_site_is_checked(client, db, search_url):
    response = client.get(search_url)
    assertContains(
        response,
        """
        <li class="list-inline-item me-3">
        <label for="id_m_0">
        <input type="radio" name="m" value="all" id="id_m_0" checked="">
        tout le site
        </label>
        </li>
        """,
        html=True,
        count=1,
    )


def test_pagination_perserves_get_params(client, db, search_url):
    forum = ForumFactory()
    topics = TopicFactory.create_batch(11, forum=forum, subject="Obtention d’un PASS IAE")
    for topic in topics:
        PostFactory(topic=topic, poster=topic.poster, subject=topic.subject)
    refresh_search_index()
    response = client.get(search_url, data={"m": "TOPIC", "q": "Obtention"})
    assertContains(
        response,
        f'<a href="{search_url}?m=TOPIC&amp;q=Obtention&amp;page=2" class="page-link">2</a>',
        count=1,
    )
