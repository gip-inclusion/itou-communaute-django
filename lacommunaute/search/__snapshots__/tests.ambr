# serializer version: 1
# name: test_extra_context[no_query]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-12">
                      <h1>Chercher dans la communauté</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="row">
                  <div class="col-12">
                      <div class="c-box">
  <form action="." class="form" method="get" name="search_form">
      <div class="form-row align-items-end">
          <div class="col">
  
  <div class="form-group" id="div_id_q">
      
          
              <label class="control-label" for="id_q">
                  Rechercher par mots clés
                  
                      <span class="text-muted">(optional)</span>
                  
              </label>
          
          
              <input class="form-control" id="id_q" maxlength="255" name="q" placeholder="Mots clés ou phrase" type="search"/>
          
      
      
      
      
  </div>
  </div>
          <div class="col-auto mb-3">
              <button class="btn btn-primary btn-ico matomo-event" data-matomo-action="search" data-matomo-category="engagement" data-matomo-option="submit_query" type="submit">
                  <i class="ri-search-line"></i>
                  <span>Rechercher</span>
              </button>
          </div>
      </div>
      <div class="form-row align-items-end">
          <div class="col-auto">Rechercher dans</div>
          <div class="col">
              <ul class="list-inline mb-0">
                  <li class="list-inline-item me-3"><label for="id_m_0"><input checked="" id="id_m_0" name="m" type="radio" value="all"/>
   tout le site</label></li><li class="list-inline-item me-3"><label for="id_m_1"><input id="id_m_1" name="m" type="radio" value="TOPIC"/>
   les échanges</label></li><li class="list-inline-item me-3"><label for="id_m_2"><input id="id_m_2" name="m" type="radio" value="FORUM"/>
   la documentation</label></li>
              </ul>
          </div>
      </div>
  </form>
  </div>
                  </div>
              </div>
              
          </div>
      </section>
  
              
          </main>
  '''
# ---
# name: test_extra_context[no_results]
  '''
  <main class="s-main" id="main" role="main">
              
                  
              
              
              
                  
      <section class="s-title-01 mt-lg-5">
          <div class="s-title-01__container container">
              <div class="s-title-01__row row">
                  <div class="s-title-01__col col-12">
                      <h1>Chercher dans la communauté</h1>
                  </div>
              </div>
          </div>
      </section>
      <section class="s-section">
          <div class="s-section__container container">
              <div class="row">
                  <div class="col-12">
                      <div class="c-box">
  <form action="." class="form" method="get" name="search_form">
      <div class="form-row align-items-end">
          <div class="col">
  
  <div class="form-group" id="div_id_q">
      
          
              <label class="control-label" for="id_q">
                  Rechercher par mots clés
                  
                      <span class="text-muted">(optional)</span>
                  
              </label>
          
          
              <input class="form-control" id="id_q" maxlength="255" name="q" placeholder="Mots clés ou phrase" type="search" value="Bubba Gump Shrimp Co."/>
          
      
      
      
      
  </div>
  </div>
          <div class="col-auto mb-3">
              <button class="btn btn-primary btn-ico matomo-event" data-matomo-action="search" data-matomo-category="engagement" data-matomo-option="submit_query" type="submit">
                  <i class="ri-search-line"></i>
                  <span>Rechercher</span>
              </button>
          </div>
      </div>
      <div class="form-row align-items-end">
          <div class="col-auto">Rechercher dans</div>
          <div class="col">
              <ul class="list-inline mb-0">
                  <li class="list-inline-item me-3"><label for="id_m_0"><input id="id_m_0" name="m" type="radio" value="all"/>
   tout le site</label></li><li class="list-inline-item me-3"><label for="id_m_1"><input checked="" id="id_m_1" name="m" type="radio" value="TOPIC"/>
   les échanges</label></li><li class="list-inline-item me-3"><label for="id_m_2"><input id="id_m_2" name="m" type="radio" value="FORUM"/>
   la documentation</label></li>
              </ul>
          </div>
      </div>
  </form>
  </div>
                  </div>
              </div>
              
                  <div class="row mt-3">
                      <div class="col-12">
                          <div class="c-box">
                              
                                  <div class="row align-items-center">
                                      <div class="col">
                                          <h3 class="h4 mb-0">Aucun résultat ? Posez votre question dans l'espace d'échanges !</h3>
                                      </div>
                                      <div class="col-12 col-md-auto mt-3 mt-md-0 d-flex align-items-center justify-content-center">
                                          <a class="btn btn-outline-primary btn-ico btn-block matomo-event" data-matomo-action="contribute" data-matomo-category="engagement" data-matomo-option="new_topic_after_search" href="/forum/forrest-gump-42/topic/create/" rel="nofollow">
                                              <i class="ri-chat-new-line ri-lg"></i>
                                              <span>Poser une question</span>
                                          </a>
                                      </div>
                                  </div>
                              
                              
                          </div>
                      </div>
                  </div>
              
          </div>
      </section>
  
              
          </main>
  '''
# ---
